Cmdline/Action/Replace.cs

```diff
@@ -0,0 +1,142 @@
+{
+    public class Replace : ICommand
+    {
+        private static readonly ILog log = LogManager.GetLogger(typeof(Replace));
+
+        public IUser User { get; set; }
+
+        public Replace(IUser user)
+        {
+            User = user;
+        }
+
+
+        public int RunCommand(CKAN.KSP ksp, object raw_options)
+        {
+            ReplaceOptions options = (ReplaceOptions) raw_options;
+
+            if (options.ckan_file != null)
+            {                
+                options.modules.Add(LoadCkanFromFile(ksp, options.ckan_file).identifier);
+            }
+
+            if (options.modules.Count == 0 && ! options.replace_all)
+            {
+                // What? No mods specified?
+                User.RaiseMessage("Usage: ckan replace Mod [Mod2, ...]");
+                User.RaiseMessage("  or   ckan replace --all");
+                return Exit.BADOPT;
+            }
+
+            User.RaiseMessage("\r\nReplacing modules...\r\n");
+            var registry = RegistryManager.Instance(ksp).registry;
+            var to_replace = new List<ModuleReplacement>();
+
+            if (options.replace_all)
+            {
+                log.Debug("Running Replace all");
+                var installed = new Dictionary<string, Version>(registry.Installed());
+
+                foreach (KeyValuePair<string, Version> mod in installed)
+                {
+                    Version current_version = mod.Value;
+
+                    if ((current_version is ProvidesVersion) || (current_version is DllVersion))
+                    {
+                        continue;
+                    }
+                    else
+                    {
+                        try
+                        {
+                            log.DebugFormat("Testing {0} {1} for possible replacement", mod.Key, mod.Value);
+                            // Check if replacement is available
+
+                            ModuleReplacement replacement = registry.GetReplacement(mod.Key, ksp.VersionCriteria());
+                            if (replacement != null)
+                            {
+                                // Replaceable
+                                log.InfoFormat("Replacement {0} {1} found for {2} {3}",
+                                    replacement.ReplaceWith.identifier, replacement.ReplaceWith.version,
+                                    replacement.ToReplace.identifier, replacement.ToReplace.version);
+                                to_replace.Add(replacement);
+                            }
+                        }
+                        catch (ModuleNotFoundKraken)
+                        {
+                            log.InfoFormat("{0} is installed, but it or its replacement is not in the registry",
+                                mod.Key);
+                        }
+                    }
+                }
+            }
+            else
+            {
+                foreach (string mod in options.modules)
```

Comment:
Currently if I try to replace mods that aren't replaceable...

```
$ mono ckan.exe replace KerbalEngineerRedux

Replacing modules...


Done!
```

... there's no error message, it just says that it finished. Similarly, I can even give it completely invalid mod names:

```
$ mono ckan.exe replace ThisIsNotAValidMod

Replacing modules...


Done!
```

If I'm trying to replace a mod that actually is replaceable but I mistype the name or capitalize it wrong, I might think I have just performed the replacement when in fact I have done nothing. I think it would be good to raise errors here if mods the user specified are invalid or not replaceable.

Reply:
All those sort of issues are logInfos. By default, the CLI only reports actual actions it takes.

Reply:
No, the CLI reports errors when mods are missing:

```
$ mono ckan install ThisIsNotAValidMod
Module ThisIsNotAValidMod required but it is not listed in the index, or not available for your version of KSP.
If you're lucky, you can do a `ckan update` and try again.
Try `ckan install --no-recommends` to skip installation of recommended modules.
```